hands-on-exercise/1-native/4-add-message.md

```diff
@@ -0,0 +1,380 @@
+
+import "cosmos/msg/v1/msg.proto";
+import "gogoproto/gogo.proto";
+import "amino/amino.proto";
+import "alice/checkers/v1/types.proto";
+import "cosmos_proto/cosmos.proto";
+
+// Msg defines the module Msg service.
+service Msg {
+  option (cosmos.msg.v1.service) = true;
+
+  // CreateGame create a game.
+  rpc CreateGame(MsgCreateGame)
+    returns (MsgCreateGameResponse);
+}
+
+// MsgCreateGame defines the Msg/CreateGame request type.
+message MsgCreateGame {
+  option (cosmos.msg.v1.signer) = "creator";
+  option (amino.name) = "alice/checkers/MsgCreateGame";
+
+  // creator is the message sender.
+  string creator = 1;
+  string index = 2 ;
+  string black = 3 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  string red = 4 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+}
+
+// MsgCreateGameResponse defines the Msg/CreateGame response type.
+message MsgCreateGameResponse {}
+```
+
+<HighlightBox type="note">
+
+* `MsgCreateGame` does not mention `Board` or `Turn` as this (as mentioned) should not be under the control of the message sender.
+* The response is empty as there is no extra information to return. For instance, here the game index is known in advance.
+* `option (cosmos.msg.v1.signer)` identifies `creator` as the field that will serve as the signer. At compilation time, the SDK will automagically pick up the value of this annotation to have `MsgCreateGame` implement `sdk.Msg.GetSigners`.
+
+</HighlightBox>
+
+### Compile Protobuf
+
+Since you have defined a new message type and an associated `service`, you should recompile everything:
+
+```sh
+$ make proto-gen
+```
+
+In the new `tx.pg.go`, you can find `type MsgCreateGame struct` and `type MsgServer interface`. Now you can use them closer to the keeper.
+
+### New message server
+
+Create a new file `keeper/msg_server.go` and take inspiration from `minimal-module-example`. It needs to:
+
+* Check that the `index` is not already taken.
+* Create a new board with the game rules.
+* If valid, put the game into storage.
+
+```go [keeper/msg_server.go]
+package keeper
+
+import (
+    "context"
+    "errors"
+    "fmt"
+
+    "cosmossdk.io/collections"
+    "github.com/alice/checkers"
+    "github.com/alice/checkers/rules"
+)
+
+type msgServer struct {
+    k Keeper
+}
+
+var _ checkers.MsgServer = msgServer{}
+
+// NewMsgServerImpl returns an implementation of the module MsgServer interface.
+func NewMsgServerImpl(keeper Keeper) checkers.MsgServer {
+    return &msgServer{k: keeper}
+}
+
+// CreateGame defines the handler for the MsgCreateGame message.
+func (ms msgServer) CreateGame(ctx context.Context, msg *checkers.MsgCreateGame) (*checkers.MsgCreateGameResponse, error) {
+    if _, err := ms.k.StoredGameList.Get(ctx, msg.Index); err == nil || errors.Is(err, collections.ErrEncoding) {
+        return nil, fmt.Errorf("game already exists at index: %s", msg.Index)
+    }
+
+    newBoard := rules.New()
+    storedGame := checkers.StoredGame{
+        Index: msg.Index,
```

Comment:
since this is a string it could be unbounded, we have had reports of this causing issues on chains. Could we make this index an integer or put a limit on the total size of the accepted amount?

Reply:
I will look into it.